.github/workflows/main.yaml

```diff
@@ -0,0 +1,170 @@
+        if: steps.changed-files.outputs.any_changed == 'true'
+        id: meta
+        uses: docker/metadata-action@v4
+        with:
+          images: ${{ env.REGISTRY }}/${{ github.repository_owner }}/cobalt-${{ env.docker_service }}
+          tags: |
+            type=ref,event=branch
+            type=ref,event=tag
+            type=ref,event=pr
+      - name: Build containers with docker-compose
+        if: steps.changed-files.outputs.any_changed == 'true'
+        uses: smartlyio/docker-compose-action@v1
+        with:
+          serviceName: ${{ env.docker_service }}
+          build: false
+          push: "on:push"
+          composeArguments: "--no-start"
+      - name: Tag images
+        if: steps.changed-files.outputs.any_changed == 'true'
+        run: |
+          docker tag cobalt-${{ env.docker_service }} ${{ steps.meta.outputs.tags }}
+          echo "DOCKER_TAG=${{steps.meta.outputs.tags}}" >> $GITHUB_ENV
+      - name: Push images
+        if: ${{ (steps.changed-files.outputs.any_changed == 'true') }}
+        run: docker push ${{ steps.meta.outputs.tags }}
+      - name: Set Docker Tag
+        if: ${{ (steps.changed-files.outputs.any_changed == 'false') && (github.event_name == 'pull_request') }}
+        run: echo "DOCKER_TAG=ghcr.io/${{github.repository_owner}}/cobalt-${docker_service}:${GITHUB_BASE_REF}" >> $GITHUB_ENV
+      - name: Set Docker Tag
+        if: ${{ (steps.changed-files.outputs.any_changed == 'false') && (github.event_name != 'pull_request') }}
+        run: echo "DOCKER_TAG=ghcr.io/${{github.repository_owner}}/cobalt-${docker_service}:${GITHUB_REF_NAME}" >> $GITHUB_ENV
+      - name: Show Docker Tag
+        run: echo $DOCKER_TAG
+    outputs:
+      docker_tag: ${{ env.DOCKER_TAG }}
+
+  # Generates build matrix based on json configuration file.
+  build-matrix:
+    runs-on: ubuntu-latest
+    steps:
+      - id: checkout
+        uses: actions/checkout@v3
+      - id: set-platforms
+        run: echo "::set-output name=platforms::$(cat ${GITHUB_WORKSPACE}/.github/config/${{ inputs.platform }}.json | jq -c '.platforms')"
+      - id: set-includes
+        run: echo "::set-output name=includes::$(cat ${GITHUB_WORKSPACE}/.github/config/${{ inputs.platform }}.json | jq -c '.includes')"
+    outputs:
+      platforms: ${{ steps.set-platforms.outputs.platforms }}
+      includes: ${{ steps.set-includes.outputs.includes }}
+  # Runs Cobalt CI workflow.
+  build:
+    needs: [docker-image, build-matrix]
+    runs-on: [self-hosted, linux, X64]
+    name: ${{matrix.platform}}_${{matrix.config}}
+    strategy:
+      fail-fast: false
+      matrix:
+        platform: ${{ fromJson(needs.build-matrix.outputs.platforms) }}
+        include: ${{ fromJson(needs.build-matrix.outputs.includes) }}
+        config: [devel, debug, qa, gold]
+    container: ${{ needs.docker-image.outputs.docker_tag }}
+    steps:
+      - name: checkout
+        uses: actions/checkout@v2
+      - name: Set env vars
```

Comment:
Why do we have to do this here instead of the general env variable setup?

Reply:
Apparently, GitHub CI didn't allow setting env var dynamically like so:

env:
  PYTHONPATH=$SOME_VARIABLE

I think it's a security precaution.

Otherwise, I moved STARBOARD_TOOLCHAINS_DIR to env section.